Propose a new architecture with focus on scalability and robustness (#360)
kubeflow/mpi-operator: proposals/scalable-robust-operator.md

3. If configured, it creates a Volcano PodGroup
4. Creates a StatefulSet for workers (plain pods in v1). The Pod template
  includes:
  - mounts for the ConfigMap.
  - `sleep` as command.
5. Creates launcher Job (plain pod in v1). The Pod template includes:
  - An init container, `kubectl-delivery`, described below.
  - Environment variables for:
    - replacing `ssh` for `kubexec.sh`
    - the `hostfile` location
  - Volumes for:
    - The ConfigMap
    - Sharing files from `kubectl-delivery`

The launcher Job, as previously mentioned, contains an init container:
`kubectl-delivery`. This is a Kubernetes controller that watches pods. It does
the following:

1. Copy kubectl from the image into the volume shared with the main container.
2. Wait for all Pods in the hostfile to be running
3. Generates a file mapping pod name to IP, in `/etc/hosts` format.

To update the status of an MPIJob, the controller uses the status of the
launcher Job. That is, when the launcher Job fails or succeeds, it’s status is
copied to the MPIJob. In v1, it bases the status on the termination condition of
the Pod.

### Analysis

The above architecture for MPI Jobs has the following problems:

> - It doesn’t scale as we increase the number of workers, belonging to the same

Reviewer: I believe this issue is addressed in v1 controller when we enable horovod elastic work on mpi-operator. https://github.com/kubeflow/mpi-operator/blob/master/proposals/elastic-horovod.md

Besides, I'm not sure what it means 'workers belonging to the same or different jobs'. What is the user case to increase the number of workers for a *different* job?

Reviewer 2: The sentence perhaps needs some rephrasing, I think the point here is that the current setup doesn't scale when increasing the number of workers of a job (e.g., running a 1000 worker pod Job), or increasing the number of jobs (running 100s of MPI jobs). 

The current design puts a lot pressure on central component, the api server:

- `kubectl exec` runs through the API server, and so all jobs running on the same cluster will have to launch their workers using connections established through the api server, which is not scalable. This doesn't scale as we increase the total number of workers across all jobs running on the same cluster.
- pod cache syncs are expensive on large clusters, doing that for each job results in streaming a large amount of state from the api server to the launcher pod. This doesn't scale as we increase the number of jobs.

  or different jobs.
  - Due to the use of `kubectl exec`, every worker spawn goes through
    kube-apiserver. `mpirun` starts a daemon in each worker
    (like [`orted`](https://www.open-mpi.org/doc/v3.0/man1/orted.1.php)).
    This process handles the worker-to-worker communication, which happens
    without the intervention of kube-apiserver. However, the `exec` connection
    stays up for control during the entirety of the job.
  - The `kubectl-delivery` controller does a full cache sync to be able to watch
    Pods. This startup penalty increases with the number of pods in the cluster

>     and has to be paid for every job. The API calls also causing additional

Reviewer: ```suggestion
    and has to be paid for every job. The API calls also cause additional
```

    stress on the apiserver as the number of launchers increases.
  - The launcher pod has execution permissions on any other Pod in the

>     namespace, which can be a security concern.

Reviewer 2: the scale issue here is that to workaround the security concern, each job needs its own namespace, which is not scalable.

Author: According to the other comment, this is not a security issue, as the role has the list of pods. But it's still a scalability issue. See update.

>     namespace, which can be a security concern.

Reviewer 3: The launcher has only the execution permission on the worker pods belong it's job, you can find it at [code](https://github.com/kubeflow/mpi-operator/blob/master/pkg/controllers/v1/mpi_job_controller.go#L1222)

Author: Ah, missed that. However, that doesn't scale as the number of workers increase (k8s objects have a size limit).

Reviewer: > Ah, missed that. However, that doesn't scale as the number of workers increase (k8s objects have a size limit).

It does. We introduced this feature in supporting Horovod Elastic: https://github.com/kubeflow/mpi-operator/pull/334

Reviewer 4: If we replace kubectl exec with ssh, we can also support the elastic feature, thus I do not think it is a problem for us if I understand it correctly.

>   - The v1 controller doesn’t implement launcher pod retries, although there are

Reviewer: What are user cases for launcher pod retry?

Reviewer 2: Running jobs on a spot/preemptible VMs.

Reviewer: In my perspective, I don't think the launcher pod should be scheduled to a spot/preemptible VMs. For the worker pods, yes, because worker pods are stateless. However, the launcher pod seems a stateful instance, whose re-restart means a new job.

Author: Still, when running the workers in a preemptible VM, if any of them fail, the entire job fails, including the launcher. So we need launcher retries.

Reviewer 5: @alculquicondor What MPI implementation are you expecting users to use after launcher gets restarted to make sure the workers are aware and can reconnect to the new launcher?

Author: The process goes like this:
1. a worker terminates unexpectedly
2. the launcher notices the worker failure, it closes the rest of ssh connections and terminates with a failure
3. the launcher restarts, launches new ssh connections to the workers
This is independent of the MPI implementation.

Reviewer: > Still, when running the workers in a preemptible VM, if any of them fail, the entire job fails, including the launcher. So we need launcher retries.

From what I observed from other HPC system, like Slurm, for jobs that are not fault-tolerant, when some of the worker fails, it does not retry via a launcher restart. Instead, the entire job will be marked as failed with it resources released. The system will re-queue the failed job if 'retry' is demanded by the user and create a new job. Such process is compatible with the contemporary design of mpi-operator.

Author: OTOH, it kind of goes against the declarative and fault-tolerant approach of k8s APIs, including the Job API. Retries is what kubernetes users would expect. And if they don't need it, they could always disable it.

Reviewer 2: I think it is useful if we make automatic retries an option that we default to 0.

    plans to. So the MPIJob behaves like a plain Pod in this version.

## Design

In order to address the problems of the existing architecture, we propose the
following changes:

- **The use of `ssh` instead of `kubectl exec`.**
  
  This would avoid any pod-to-pod communication happening through apiserver and
  doesn’t require giving execution permissions at the namespace level.
  This can be achieved like this:

>   - The controller can generate a single key and share it with the launcher and

Reviewer 2: ```suggestion
  - The controller generates a single key and share it with the launcher and
```

    worker pods through a Secret.
  - The launcher and workers mount the Secret and set appropriate file
    permissions.
  - The workers run an SSH server instead of `sleep`.

>   - When encrypted communication is not a requirement, users could have the

Reviewer 2: ```suggestion
  - When encrypted communication is not a requirement, users have the
```

  - This removes the need to query Pod IPs, as the Pods can discover each other
    through DNS resolution. The hostfile can be generated statically by the
    controller using the stable hostnames.
  - Starting with k8s 1.22, we can use
    [Indexed Jobs with stable hostnames](https://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
    to delegate the pod management to Kubernetes. Additionally, this will give
    us robust failure tracking so that we can give users control over retry
    limits.
    In the meantime, we can continue using plain Pods.
    - Caveat 1: The Job controller doesn’t report the number of running Pods.
      Instead, it reports active Pods, which include running and pending
      (scheduling or starting). But Kubernetes SIG Apps is
      [open to add a status field for running Pods](https://kubernetes.slack.com/archives/C18NZM5K9/p1619549430067400).
    - Caveat 2: Horovod supports elastic workers, but the Kubernetes Job
      doesn’t support changes to the completions field. This can be supported
      starting from 1.23. In the meantime, we can replicate the behavior by
      creating a new Job and doing Pod adoption.

>   - For Intel MPI, we also need a headless Service to front the launcher.

Reviewer 2: Add: because workers connect back to the launcher.

- **Revert the use of the Job API for the launcher.**
  - The Job controller handles retries when the launcher or any of the workers fail.
  - Caveat 1 also applies: The Job controller doesn’t report if the Pod is running.
    We can continue watching Pods in the meantime.
- With the above changes, **the following objects can be removed**:
  - The ServiceAccount+Role+RoleBinding for the launcher.
  - The `kubectl-delivery` init container in the launcher, as there is no need

>     to obtain IPs, speeding up startup time.

Reviewer 3: Do u mean the init-container `kubectl-delivery` will be removed? If that, how could we  keep the launcher start after the  workers?

Author: There are 2 options:
1. Do nothing, leave such handling to the gang scheduler. For this reason it is important that the launcher can do retries.
1. To not create the launcher until all the worker pods are running. So everything is handled by the controller, no need for extra cache syncs.

Reviewer 3: I don't think so, you would like to speed up startup time, but it does not if you want the launcher to retry for failures. And the scheduler should not know about the job startup strategy.

Reviewer 3: By the same time, how do you verify the  real launcher failed reason for option 1? for example,
* workers does not startup, the launcher failed and restart
* training job failed, the launcher failed.

Author: What do you mean by "job startup strategy"?

As for differentiating failures, I don't think we can or should do that. There is another type of failure which is kind of a mix of the two you described: a worker goes down. In this case, the job might have already started.

But perhaps option 2 is reasonable. WDYT?

Reviewer 2: If we want to support retries, then by product we want option 1, correct?

Author: we can still use option 2. It might make startups less flaky. But it's not a guarantee anyways. A worker can simply just fail in between the time it reported running and the launcher started running.

      [open to add a status field for running Pods](https://kubernetes.slack.com/archives/C18NZM5K9/p1619549430067400).
    - Caveat 2: Horovod supports elastic workers, but the Kubernetes Job
      doesn’t support changes to the completions field. This can be supported
      starting from 1.23. In the meantime, we can replicate the behavior by
      creating a new Job and doing Pod adoption.
  - For Intel MPI, we also need a headless Service to front the launcher,
    because workers communicate back to the launcher using its hostname.
- **Revert the use of the Job API for the launcher.**
  - The Job controller handles retries when the launcher or any of the workers fail.
  - Caveat 1 also applies: The Job controller doesn’t report if the Pod is running.
    We can continue watching Pods in the meantime.
- With the above changes, **the following objects can be removed**:
  - The ServiceAccount+Role+RoleBinding for the launcher.
  - The `kubectl-delivery` init container in the launcher, as there is no need
    to obtain IPs, speeding up startup time.
    
## Alternatives Considered

TBD from discussions

## Appendix: Prototype objects

> It uses a StatefulSet in place of Indexed Jobs, as they are still an alpha

Reviewer 5: Could you add a link to "Indexed Jobs"?

```

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: mpi-launcher
spec:
  template:
    spec:
      restartPolicy: OnFailure
      containers:
      - name: driver
        image: '<USER_IMAGE>'
        args:
        - 'mpirun'
        - '-np'
        - '9'
        - '<USER_EXECUTABLE>'
        env:
        - name: 'OMPI_MCA_orte_keep_fqdn_hostnames'
          value: 'true'
        - name: 'OMPI_MCA_orte_default_hostfile'
          value: '/home/mpiuser/config/hostfile'
        volumeMounts:
        - name: ssh-auth
          mountPath: /mnt/ssh
          readOnly: true
        - name: mpi-config
          mountPath: /home/mpiuser/config
          readOnly: true
      volumes:
      - name: mpi-config
        configMap:
          name: mpi-config
      - name: ssh-auth
        secret:
          secretName: ssh-auth
          items:
          - key: ssh-privatekey
            path: id_rsa
          - key: ssh-publickey
            path: id_rsa.pub
          - key: ssh-publickey
            path: authorized_keys
```

```yaml
apiVersion: v1
kind: Service
metadata:
  name: mpi-workers
spec:
  clusterIP: None
  selector:
    app: mpi-workers
```

```yaml
apiVersion: apps/v1

> kind: StatefulSet

Reviewer 4: I am wondering how to support elastic mode with statefulset? We may add/remove workers in the fly.

Author: You just change `.spec.replicas`.
